packages/library-base/src/commonMain/kotlin/io/realm/internal/BaseResults.kt

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal
+
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.interop.Callback
+import io.realm.internal.interop.Link
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.RealmCoreException
+import io.realm.internal.interop.RealmInterop
+import kotlinx.coroutines.channels.ChannelResult
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+internal abstract class BaseResults<E : Any> constructor(
+    protected val realm: RealmReference,
+    internal val nativePointer: NativePointer,
+    protected val clazz: KClass<E>,
+    protected val mediator: Mediator,
+    protected val mode: Mode = Mode.RESULTS
+) : AbstractList<E>(), RealmResults<E>, Freezable<BaseResults<E>>, Thawable<BaseResults<E>>, Observable<BaseResults<E>>, RealmStateHolder {
+
+    enum class Mode {
+        // FIXME Needed to make working with @LinkingObjects easier.
+        EMPTY, // RealmResults that is always empty.
+        RESULTS // RealmResults wrapping a Realm Core Results.
+    }
+
+    abstract fun instantiateResults(
+        realmReference: RealmReference,
+        nativePointer: NativePointer,
+        clazz: KClass<E>,
+        mediator: Mediator
+    ): BaseResults<E>
+
+    override val size: Int
+        get() = RealmInterop.realm_results_count(nativePointer).toInt()
+
+    @Suppress("SpreadOperator")
+    override fun query(query: String, vararg args: Any?): BaseResults<E> {
+        try {
+            val table = clazz.simpleName!!
+            val queryPointer = RealmInterop.realm_query_parse(nativePointer, table, query, *args)
+            val resultsPointer = RealmInterop.realm_query_find_all(queryPointer)
```

Comment:
Not 100% sure if this method is just for the old API 🤔. If yes, just ignore this comment.

If not, calling `find_all` here should be avoided as it evaluates the query directly on the calling thread. Results in ObjectStore has a constructor that takes the Query object directly.